Only add primary column when needed.

peterwilsoncc opened the pull request:
https://core.trac.wordpress.org/ticket/34564

peterwilsoncc committed:
Only add primary column when needed.

peterwilsoncc committed:
This is safer.

peterwilsoncc committed:
Only Mx Forty forgot to invite you.

peterwilsoncc committed:
Merge branch 'master' into 34564-column-info-performance

peterwilsoncc committed:
Merge branch 'trunk' into 34564-column-info-performance

spacedmonkey commented on src/wp-admin/includes/class-wp-list-table.php:
```diff
@@ -1108,20 +1108,29 @@ protected function get_primary_column_name() {
 			 * column headers property. This ensures the primary column name is included
 			 * in plugins setting the property directly in the three item format.
 			 */
+			if ( 4 === count( $this->_column_headers ) ) {
+				return $this->_column_headers;
+			}
```
What is happening here?

mukeshpanchal27 reviewed (approved):
Looks great!

peterwilsoncc committed:
Merge branch 'trunk' into 34564-column-info-performance

peterwilsoncc committed:
Merge branch 'trunk' into 34564-column-info-performance

peterwilsoncc committed:
Tests for list tables. props @costdev

peterwilsoncc committed:
Merge branch 'trunk' into 34564-column-info-performance

peterwilsoncc committed:
Remove tests for `WP_Post_Comments_List_Table`.

Comment headers are hard coded in a function call rather than the `_column_headers` property thus preventing the tests from passing.

peterwilsoncc committed:
Test `list_table_primary_column` fires the expected number of times.

costdev commented on tests/phpunit/tests/admin/wpListTable.php:
```diff
@@ -0,0 +1,103 @@
+		$column_info = new ReflectionMethod( $list_table, 'get_column_info' );
+		$column_info->setAccessible( true );
+
+		$this->assertSame( $expected, $column_info->invoke( $list_table ) );
+		$this->assertSame( $expected_hook_count, $hook->get_call_count() );
```
```suggestion
		$this->assertSame( $expected, $column_info->invoke( $list_table ), 'The actual columns did not match the expected columns' );
		$this->assertSame( $expected_hook_count, $hook->get_call_count(), 'The hook was not called the expected number of times' );
```

costdev commented on tests/phpunit/tests/admin/wpListTable.php:
```diff
@@ -0,0 +1,103 @@
+	 *
+	 * @covers WP_List_Table::get_column_info
+	 *
+	 * @param string $list_class The name of the WP_List_Table child class.
+	 * @param array  $headers    A list of column headers.
+	 * @param array  $expected   The expected column headers.
```
```suggestion
	 * @param string $list_class          The name of the WP_List_Table child class.
	 * @param array  $headers             A list of column headers.
	 * @param array  $expected            The expected column headers.
	 * @param int    $expected_hook_count The expected number of times the hook is called.
```

peterwilsoncc committed:
Update tests/phpunit/tests/admin/wpListTable.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

peterwilsoncc committed:
Update tests/phpunit/tests/admin/wpListTable.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented on tests/phpunit/tests/admin/wpListTable.php:
```diff
@@ -0,0 +1,104 @@
+		$hook = new MockAction();
+		add_filter( 'list_table_primary_column', array( $hook, 'filter' ) );
+
+		// This is reset between tests.
+		$GLOBALS['hook_suffix'] = 'my-hook';
+
+		$list_table = _get_list_table( $list_class );
```
```suggestion
		$list_table = _get_list_table( $list_class, array( 'screen' => 'test-screen' ) );
```
Passing a `screen` value in the args removes the need to set the `hook_suffix` global.

peterwilsoncc committed:
Explain why the global is set in a comment.

peterwilsoncc committed:
Merge branch 'trunk' into 34564-column-info-performance

peterwilsoncc commented:
Merged https://core.trac.wordpress.org/changeset/53868 / 11cc06cd22
